src/v/cluster/node_status_provider.cc

```diff
@@ -0,0 +1,235 @@
+    if (!config::shard_local_cfg().disable_public_metrics()) {
+        setup_metrics(_public_metrics);
+    }
+
+    if (!config::shard_local_cfg().disable_metrics()) {
+        setup_metrics(_metrics);
+    }
+
+    _members_table_notification_handle
+      = _members_table.register_members_updated_notification(std::bind_front(
+        std::mem_fn(&node_status_provider::handle_members_updated_notification),
+        this));
+
+    _timer.set_callback([this] { tick(); });
+}
+
+void node_status_provider::handle_members_updated_notification(
+  std::vector<model::node_id> node_ids) {
+    for (const auto& node_id : node_ids) {
+        if (node_id == _self || _discovered_peers.contains(node_id)) {
+            continue;
+        }
+
+        vlog(
+          clusterlog.info,
+          "Node {} has been discovered via members table",
+          node_id);
+        _discovered_peers.insert(node_id);
+    }
+}
+
+void node_status_provider::start() {
+    vassert(ss::this_shard_id() == shard, "invoked on a wrong shard");
+    return tick();
+}
+
+ss::future<> node_status_provider::stop() {
+    vassert(ss::this_shard_id() == shard, "invoked on a wrong shard");
+
+    _metrics.clear();
+    _public_metrics.clear();
+
+    _members_table.unregister_members_updated_notification(
+      _members_table_notification_handle);
+    _timer.cancel();
+    return _gate.close().then([this] { return _node_connection_cache.stop(); });
+}
+
+std::optional<node_status>
+node_status_provider::get_node_status(model::node_id node) const {
+    if (node == _self) {
+        return node_status{
+          .node_id = node, .last_seen = rpc::clock_type::now()};
+    }
+
+    if (_peers_state.contains(node)) {
+        return _peers_state.at(node);
+    }
+
+    return std::nullopt;
+}
+
+void node_status_provider::update_node_status(node_status_metadata&& meta) {
+    auto [it, inserted] = _discovered_peers.insert(meta.node_id);
+    if (inserted) {
+        vlog(
+          clusterlog.info, "Node {} has been discovered via RPC request", *it);
+    }
+
+    _peers_state[meta.node_id] = {
+      .node_id = meta.node_id, .last_seen = rpc::clock_type::now()};
+}
+
+void node_status_provider::tick() {
+    ssx::background = ssx::spawn_with_gate_then(_gate, [this] {
+                          return dispatch_node_status_requests().finally(
+                            [this] {
+                                if (!_gate.is_closed()) {
+                                    _timer.arm(_period());
+                                }
+                            });
+                      }).handle_exception([](const std::exception_ptr& e) {
+        vlog(
+          clusterlog.debug,
+          "Exception during node status provider tick: {}",
+          e);
+    });
+}
+
+ss::future<> node_status_provider::dispatch_node_status_requests() {
+    node_status_request request = {.sender_metadata = {.node_id = _self}};
+
+    std::vector<ss::future<>> futures;
```

Comment:
nit: can be simplified with: `ssx::parallel_transform`

Reply:
Done.